cpp/src/io/json/read_json.cu

```diff
@@ -164,42 +166,86 @@ auto get_record_range_raw_input(host_span<std::unique_ptr<datasource>> sources,
+  size_t const total_source_size = sources_size(sources, 0, 0);
+  rmm::device_uvector<char> merged(0, stream);
+
+  rmm::device_uvector<char> cur_chunk_buf = ingest_raw_input(sources,
+                                                             reader_opts.get_compression(),
+                                                             reader_opts.get_byte_range_offset(),
+                                                             reader_opts.get_byte_range_size(),
+                                                             stream);
+  auto first_delim_pos                    = reader_opts.get_byte_range_offset() == 0
+                                              ? 0
+                                              : find_first_delimiter(cur_chunk_buf, '\n', stream);
   if (first_delim_pos == -1) {
     return rmm::device_uvector<char>{0, stream};
-  } else {
-    first_delim_pos = first_delim_pos + reader_opts.get_byte_range_offset();
+  } else if (!should_load_whole_source(reader_opts, sources[0]->size()) &&
+             cur_chunk_buf.back_element(stream) != '\n' &&
+             reader_opts.get_byte_range_offset() + reader_opts.get_byte_range_size() <
+               total_source_size) {
     // Find next delimiter
-    decltype(first_delim_pos) next_delim_pos = -1;
-    auto const total_source_size             = sources_size(sources, 0, 0);
-    auto current_offset = reader_opts.get_byte_range_offset() + reader_opts.get_byte_range_size();
-    while (current_offset < total_source_size and next_delim_pos == -1) {
-      buffer         = ingest_raw_input(sources,
-                                reader_opts.get_compression(),
-                                current_offset,
-                                reader_opts.get_byte_range_size(),
-                                stream);
-      next_delim_pos = find_first_delimiter(buffer, '\n', stream);
-      if (next_delim_pos == -1) { current_offset += reader_opts.get_byte_range_size(); }
+    /*
+     * NOTE: heuristic for choosing subchunk size: geometric mean of minimum subchunk size (set to
+     * 10kb) and the byte range size
+     */
+    std::int64_t next_delim_pos = -1;
+    constexpr int num_subchunks = 10;  // per byte_range_size
+    auto geometric_mean         = [](double a, double b) { return std::pow(a * b, 0.5); };
+    size_t size_per_subchunk =
+      geometric_mean(reader_opts.get_byte_range_size() / num_subchunks, 10000);
+    size_t next_subchunk_start =
+      reader_opts.get_byte_range_offset() + reader_opts.get_byte_range_size();
+    std::vector<rmm::device_uvector<char>> subchunk_buffers;
+
+    while (next_subchunk_start < total_source_size && next_delim_pos == -1) {
+      subchunk_buffers.emplace_back(ingest_raw_input(
+        sources, reader_opts.get_compression(), next_subchunk_start, size_per_subchunk, stream));
+      next_delim_pos = find_first_delimiter(subchunk_buffers.back(), '\n', stream);
```

Comment:
maybe shrink the last subchunk if the delimiter is found? should be practically free, and might simplify some of the logic below (e.g. always copy whole vector)

Reply:
That's a good idea - we will reduce memory cost there. But if we `shrink_to_fit`, wouldn't that be expensive since we have to do it on the same passed stream? I couldn't think of a cleaner way to include the logic for the next delimiter not being found (lines 206-207).